### PR TITLE
Permalink and API-Index for markdown output.

### PR DESCRIPTION
**Pull request by peter1000**
* Add optional API-Index for markdown output.
* Add optional Permalink for markdown output.
* Add an example MkDocs theme: `mkdocshelptheme` to show how to use the
`bootstrap` related themes.
* Add some additional explanatory code comments / documentation.
* Adjusted some documentation for the new API-Index. (Shorter first line
ect..)

Issue: https://github.com/MichaelHatherly/Lexicon.jl/issues/43
replaces PR: https://github.com/MichaelHatherly/Lexicon.jl/pull/48

### CODE REVIEW

File: NEWS.md [hunk 1]
```diff
@@ -1,3 +1,11 @@
+### v0.1.5dev
+
+* Add optional API-Index for markdown output. Contributed by [@peter1000](https://github.com/peter1000).
+* Add optional Permalink for markdown output. Contributed by [@peter1000](https://github.com/peter1000).
+* Add an example MkDocs theme: `mkdocshelptheme` to show how to use the `bootstrap` related themes.
+* Add some additional explanatory code comments / documentation.
+* Adjusted some documentation for the new API-Index. (Shorter first line ect..)
```
Comment (MichaelHatherly) [inline NEWS.md]: "ect.." -> "etc."

File: NEWS.md [hunk 2]
```diff
@@ -17,7 +25,7 @@
 
 ## v0.1.0
 
-* Add matkdown output.
+* Add markdown output.
```
Comment (MichaelHatherly) [inline NEWS.md]: Thanks.

File: docs/build.jl [hunk 3]
```diff
@@ -3,6 +3,17 @@ using Docile, Lexicon
 const api_directory = "api"
 const modules = [Lexicon]
 
+"""
+Example of customary MDSTYLE
```
Comment (MichaelHatherly) [inline docs/build.jl]: "customary" -> "custom"

File: docs/build.jl [hunk 4]
```diff
@@ -16,15 +27,19 @@ cd(dirname(@__FILE__)) do
...
         try
-            save(filename, m)
+            save(filename, m, plmain)
         catch err
             println(err)
             exit(1)
         end
     end
+    # To actually save the API Index to: `genindex.md`: If not index page is needed this can be skipped
+    savegenidx(plmain)
```
Comment (MichaelHatherly) [inline docs/build.jl]: Could the API be streamlined a bit here. Pass in an index object which is then mutated and then saving the result seems too complex to me.

How about having `save` return an index object and them applying `save` the to resulting indices vector?

Kind of like:

```julia
indices = []
for m in modules
    filename = joinpath(api_directory, "$(module_name(m)).md")
    push!(indices, save(filename, m))
end
save(joinpath(api_directory, "genindex.md"), indices; headerstyle = "#", modnamestyle = "##")
```

I've not investigated whether that's feasible yet, so please do say if it isn't.

File: docs/mkdocshelptheme/README.md [hunk 5]
```diff
@@ -0,0 +1,29 @@
+MkDocs related example to fix the permalink below the top bannel.
```
Comment (MichaelHatherly) [inline docs/mkdocshelptheme/README.md]: "bannel" -> "banner"

File: docs/mkdocshelptheme/README.md [hunk 6]
```diff
@@ -0,0 +1,29 @@
...
+pages:
+    - ['index.md', 'Introduction']
+    - ['manual.md', 'Manual', 'Overview']
+    - ['api/Lexicon.md', 'API', 'Lexicon']
+    - ['api/genindex.md', 'API', 'Index']
+
+theme:               bootstrap
+# Example to fix the permalink below the top bannel
+theme_dir:           docs/mkdocshelptheme   
+```
```
Comment (MichaelHatherly) [inline docs/mkdocshelptheme/README.md]: Needs an extra blank line after `theme_dir: ...` to skip `doctest`.

File: docs/mkdocshelptheme/README.md [hunk 7]
```diff
@@ -0,0 +1,29 @@
...
+
+and re-run:
+
+```
+$ mkdocs build --clean
+```
```
Comment (MichaelHatherly) [inline docs/mkdocshelptheme/README.md]: When testing I typically just use `mkdocs serve` to  here to avoid generating files.

File: src/query.jl [hunk 8]
```diff
@@ -177,8 +178,8 @@ function partial_signature_matching(f::Function, sig::Tuple)
 end
 
 """
-Search loaded documentation for all methods of a generic function `f` that match
-the provided signature `sig`. Optionally, provide an index (1-based) to view an
+Search for all methods of a generic function `f` that match the signature `sig`.
+Optionally, provide an index (1-based) to view an
```
Comment (MichaelHatherly) [inline src/query.jl]: The "loaded documentation" is important here since users might assume that Lexicon will search through all docstrings even if their respective module hasn't been `using ModuleName` yet.

File: src/render.jl [hunk 9]
```diff
@@ -98,15 +143,25 @@ The documentation will be available from
 `https://USER_NAME.github.io/PACKAGE_NAME/FILE_PATH.html`.
 
 """
-function save(file::String, modulename::Module; mathjax = false, include_internal = true)
+function save(file::String, modulename::Module, plmain;
+                            mathjax = false, include_internal = true, permalink::Bool = true)
     mime = MIME("text/$(strip(last(splitext(file)), '.'))")
-    save(file, mime, documentation(modulename); mathjax = mathjax, include_internal = include_internal)
+    save(file, mime, documentation(modulename), plmain; mathjax = mathjax,
+        include_internal = include_internal, permalink = permalink)
```
Comment (MichaelHatherly) [inline src/render.jl]: I think we're going to need a `Dict{Symbol, Any}`, or something, with the potential increase in configuration settings.

```julia
function save(file::String, modulename::Module, plmain, config::Dict{Symbol, Any})
    mime = MIME("text/$(strip(last(splitext(file)), '.'))")
    save(file, mime, documentation(modulename), plmain, config)
end
```

Not with this PR though, we can change it later on.

File: src/render/md.jl [hunk 10]
```diff
@@ -30,14 +31,184 @@ function validate(mdstyle::Dict{Symbol, ByteString})
...
+
+"Permalink Main helper type."
+type PLMain
+    genidxfile::ByteString
+    mdoutfile::ByteString      # reference to the current proccessed output file needed for relative links.
+    idxrelpath::ByteString     # current relative path: mdoutfile to genidxfile
+    headerstyle::ByteString
+    modnamestyle::ByteString
+    data::Dict{ByteString, Dict{Symbol, Array{PLEntry,1}}}
+end
```
Comment (MichaelHatherly) [inline src/render/md.jl]: Could we have more direct names for `PLEntry` and `PLMain`. Perhaps `Permalink` and `Index`?

File: src/render/md.jl [hunk 11]
```diff
@@ -30,14 +31,184 @@ function validate(mdstyle::Dict{Symbol, ByteString})
...
+        path_arr = split(path, Base.path_separator_re)
+    else
+        path_arr = split(abspath(path), Base.path_separator_re)
+    end
+    if isabspath(startpath) && !contains(path, "/.")
+        start_arr = split(startpath, Base.path_separator_re)
+    else
+        start_arr = split(abspath(startpath), Base.path_separator_re)
+    end
+
```
Comment (MichaelHatherly) [inline src/render/md.jl]: Could be shortened to:

```julia
path      = isabspath(path) &&      !contains(path, "/.") ? path      : abspath(path)
startpath = isabspath(startpath) && !contains(path, "/.") ? startpath : abspath(startpath)
path_arr  = split(path,      Base.path_separator_re)
start_arr = split(startpath, Base.path_separator_re)
```

File: src/render/md.jl [hunk 12]
```diff
@@ -30,14 +31,184 @@ function validate(mdstyle::Dict{Symbol, ByteString})
...
+        count += 1
+        if path_arr[count] != start_arr[count]
+            count -= 1
+            break
+        end
+    end
+    rel_path = "..$(Base.path_separator)"^(length(start_arr)-count) * join(path_arr[count+1:end],
+                                                                             Base.path_separator)
+    return rel_path
+end
```
Comment (MichaelHatherly) [inline src/render/md.jl]: I assume this method works on Windows (don't have one to try it on)? Tests for it would be good to have.

File: src/render/md.jl [hunk 13]
```diff
@@ -30,14 +31,184 @@ function validate(mdstyle::Dict{Symbol, ByteString})
...
+
+"""
+Starts collecting data for API-Index.
+
+This must be called before any Modules are saved. Even in the case one does not want to save
+a API-Index page.
```
Comment (MichaelHatherly) [inline src/render/md.jl]: See comment further up about streamlining the API a bit.

File: src/render/md.jl [hunk 14]
```diff
@@ -30,14 +31,184 @@ function validate(mdstyle::Dict{Symbol, ByteString})
...
+    headerstyle in vaild_tags || error("Invalid headerstyle:  `$headerstyle`. Valid values: $vaild_tags_str.")
+    modnamestyle in vaild_tags || error("Invalid modnamestyle:  `$modnamestyle`. Valid values: $vaild_tags_str.")
+    return PLMain(abspath(genidxfile), "", "", headerstyle, modnamestyle, Dict())
+end
+
+"""
+Saves the generated API-Index to file.
+
+This must be called before any Modules are saved. Even in the case one does not want to save
+a API-Index page.
```
Comment (MichaelHatherly) [inline src/render/md.jl]: Is this an unintended duplication of the sentence from the docstring above?

File: src/render/md.jl [hunk 15]
```diff
@@ -30,14 +31,184 @@ function validate(mdstyle::Dict{Symbol, ByteString})
...
+"""
+Prepares a PLEntry - pushes it to the correct `plmain` Vector.
+It adds any necessary dictionary keys.
+
+**Returns:**
+
+* PLEntry: Returns also the just prepared/added PLEntry.
+"""
+function push!(plmain::PLMain, modname::Module, obj, ent::Entry)
+    isa(ent, Docile.Entry{:macro}) ? basename = Docile.Interface.macroname(ent) : basename = Docile.Interface.name(obj)
```
Comment (MichaelHatherly) [inline src/render/md.jl]: Bit shorter:
```julia
basename = isa(ent, Docile.Entry{:macro}) ? Docile.Interface.macroname(ent) : Docile.Interface.name(obj)
```

File: src/render/md.jl [hunk 16]
```diff
@@ -30,14 +31,184 @@ function validate(mdstyle::Dict{Symbol, ByteString})
...
+    modname = string(modname)
+    relsourcepath = relpath(plmain.mdoutfile, dirname(plmain.genidxfile))
+    plentry = PLEntry(writeobj(obj, ent), "", split(docs(ent).data, '\n')[1], relsourcepath)
```
Comment (MichaelHatherly) [inline src/render/md.jl]: `docs(ent).data` -> `data(docs(ent))`

File: src/render/md.jl [hunk 17]
```diff
@@ -30,14 +31,184 @@ function validate(mdstyle::Dict{Symbol, ByteString})
...
+
+    if modname in keys(plmain.data)
+        if basename in keys(plmain.data[modname])
+            idnum = length(plmain.data[modname][basename]) + 1
+            plentry.idname = "$basename.$idnum"
+            temp_arr = plmain.data[modname][basename]
+            Base.push!(temp_arr, plentry)
```
Comment (MichaelHatherly) [inline src/render/md.jl]: Is `temp_arr` redundant here?

File: src/render/md.jl [hunk 18]
```diff
@@ -30,14 +31,184 @@ function validate(mdstyle::Dict{Symbol, ByteString})
...
+        plmain.data[modname] = Dict([(basename, [plentry])])
+    end
+    return plentry
+end
+
+function save(file::String, mime::MIME"text/md", doc::Metadata, plmain::PLMain,
+                mdstyle::Dict{Symbol, ByteString} = DEFAULT_MDSTYLE;
+                mathjax = false, include_internal = true, permalink::Bool = true)
     validate(mdstyle)
+    # update plmain with the: mdoutfile needed to calculate the relative Api-Index links.
```
Comment (MichaelHatherly) [inline src/render/md.jl]: "update" -> "Update"

"the:" -> "the"

**Comment by MichaelHatherly** [comment]
Added notes inline. Only real concern is the api that we settle on. I'd like to avoid users having to worry about creating an index when they don't want one.

File: src/render/md.jl (hunk 12 above)
Comment (peter1000) [inline src/render/md.jl]: I have also only access to Linux - so I'm not sure if it Works on Windows

**Comment by peter1000** [comment]
> Only real concern is the api that we settle on. I'd like to avoid users having to worry about creating an index when they don't want one. 

thanks for the comments and helpful input - also sorry English is not my first language.

anyway, I will look into your suggestion - probably today or tomorrow.

File: docs/mkdocshelptheme/README.md (hunk 5 above)
Comment (peter1000) [inline docs/mkdocshelptheme/README.md]: mkdocshelptheme folder will be removed: checked with MkDocs team for a better solution

File: docs/mkdocshelptheme/README.md (hunk 6 above)
Comment (peter1000) [inline docs/mkdocshelptheme/README.md]: mkdocshelptheme folder will be removed: checked with MkDocs team for a better solution

File: docs/mkdocshelptheme/README.md (hunk 7 above)
Comment (peter1000) [inline docs/mkdocshelptheme/README.md]: mkdocshelptheme folder will be removed: checked with MkDocs team for a better solution

File: src/query.jl (hunk 8 above)
Comment (peter1000) [inline src/query.jl]: Can I do it like this:

```
Search loaded documentation for all methods of a generic function `f` 
that match the provided signature `sig`.
```
So that for the index describtion the first line is not too long - or better the whole sentence in the first line?

File: NEWS.md (hunk 1 above)
Comment (peter1000) [inline NEWS.md]: Fixed

File: docs/build.jl (hunk 3 above)
Comment (peter1000) [inline docs/build.jl]: Fixed

File: src/render/md.jl (hunk 11 above)
Comment (peter1000) [inline src/render/md.jl]: Fixed

File: src/render/md.jl (hunk 15 above)
Comment (peter1000) [inline src/render/md.jl]: Fixed

File: src/render/md.jl (hunk 17 above)
Comment (peter1000) [inline src/render/md.jl]: Fixed

**Comment by peter1000** [comment]
Added some notes inline. I will continue to push to the same branch and later rebase for a clean PR.

About streamlining the API - I will have to rethink a bit....

I also opened a PR at MkDocs which would make it easier to use the permalinks with the bootstrap themes. It would be greate if they could include this in there upcoming v.012 release. See https://github.com/mkdocs/mkdocs/pull/448

File: src/render/md.jl (hunk 12 above)
Comment (MichaelHatherly) [inline src/render/md.jl]: Probably just a new `context` in the `tests/rendering.jl` file for now. Something like:

```julia
context("Relative Paths.") do
    @fact relpath("<path one>", "<path two>") => "<expected result>"
    # and a few more, perhaps any edge cases such as empty paths...
end
```

File: src/render.jl (hunk 9 above)
Comment (peter1000) [inline src/render.jl]: Opened a new issue so it want be forgotten.

**Comment by peter1000** [comment]
With the many good changes so far I think there is not point to keep this PR. I will do later a new one (probably 2 split the permalink and the index page generation)
